src/poisson_fft.f90

```diff
@@ -0,0 +1,198 @@
+      implicit none
+
+      class(poisson_fft_t) :: self
+      type(dirps_t), intent(in) :: xdirps, ydirps, zdirps
+      integer, intent(in) :: sz
+
+      complex(dp), allocatable, dimension(:) :: xkx, xk2, yky, yk2, zkz, zk2, &
+                                                exs, eys, ezs
+
+      integer :: nx, ny, nz
+      real(dp) :: w, wp, rlexs, rleys, rlezs, xtt, ytt, ztt, xt1, yt1, zt1
+      complex(dp) :: xt2, yt2, zt2, xyzk
+
+      integer :: i, j, ka, kb, ix, iy, iz
+
+
+      nx = xdirps%n; ny = ydirps%n; nz = zdirps%n
+
+      do i = 1, nx
+         self%ax(i) = sin((i-1)*pi/nx)
+         self%bx(i) = cos((i-1)*pi/nx)
+      end do
+
+      do i = 1, ny
+         self%ay(i) = sin((i-1)*pi/ny)
+         self%by(i) = cos((i-1)*pi/ny)
+      end do
+
+      do i = 1, nz
+         self%az(i) = sin((i-1)*pi/nz)
+         self%bz(i) = cos((i-1)*pi/nz)
+      end do
+
+      ! Now kxyz
+      allocate(xkx(nx), xk2(nx), exs(nx))
+      allocate(yky(ny), yk2(ny), eys(ny))
+      allocate(zkz(nz), zk2(nz), ezs(nz))
+      xkx(:) = 0; xk2(:) = 0; yky(:) = 0; yk2(:) = 0; zkz(:) = 0; zk2(:) = 0
+
+      ! periodic-x
+      do i = 1, nx/2 + 1
+         w = 2*pi*(i - 1)/nx
+         wp = xdirps%stagder_v2p%a*2*xdirps%d*sin(0.5_dp*w) &
+              + xdirps%stagder_v2p%b*2*xdirps%d*sin(1.5_dp*w)
+         wp = wp/(1._dp + 2*xdirps%stagder_v2p%alpha*cos(w))
+
+         xkx(i) = cmplx(1._dp, 1._dp, kind=dp)*(nx*wp/xdirps%L)
+         exs(i) = cmplx(1._dp, 1._dp, kind=dp)*(nx*w/xdirps%L)
+         xk2(i) = cmplx(1._dp, 1._dp, kind=dp)*(nx*wp/xdirps%L)**2
+      end do
+      do i = nx/2 + 2, nx
+         xkx(i) = xkx(nx - i + 2)
+         exs(i) = exs(nx - i + 2)
+         xk2(i) = xk2(nx - i + 2)
+      end do
+
+      ! periodic-y
+      do i = 1, ny/2 + 1
+         w = 2*pi*(i - 1)/ny
+         wp = ydirps%stagder_v2p%a*2*ydirps%d*sin(0.5_dp*w) &
+              + ydirps%stagder_v2p%b*2*ydirps%d*sin(1.5_dp*w)
+         wp = wp/(1._dp + 2*ydirps%stagder_v2p%alpha*cos(w))
+
+         yky(i) = cmplx(1._dp, 1._dp, kind=dp)*(ny*wp/ydirps%L)
+         eys(i) = cmplx(1._dp, 1._dp, kind=dp)*(ny*w/ydirps%L)
+         yk2(i) = cmplx(1._dp, 1._dp, kind=dp)*(ny*wp/ydirps%L)**2
+      end do
+      do i = ny/2 + 2, ny
+         yky(i) = yky(ny-i+2)
+         eys(i) = eys(ny-i+2)
+         yk2(i) = yk2(ny-i+2)
+      end do
+
+      ! periodic-z
+      do i = 1, nz/2 + 1
+         w = 2*pi*(i - 1)/nz
+         wp = zdirps%stagder_v2p%a*2*zdirps%d*sin(0.5_dp*w) &
+              + zdirps%stagder_v2p%b*2*zdirps%d*sin(1.5_dp*w)
+         wp = wp/(1._dp + 2*zdirps%stagder_v2p%alpha*cos(w))
+
+         zkz(i) = cmplx(1._dp, 1._dp, kind=dp)*(nz*wp/zdirps%L)
+         ezs(i) = cmplx(1._dp, 1._dp, kind=dp)*(nz*w/zdirps%L)
+         zk2(i) = cmplx(1._dp, 1._dp, kind=dp)*(nz*wp/zdirps%L)**2
+      end do
+
+      print*, 'set waves array'
+      ! TODO: do loop ranges below are valid only for single rank runs
```

Comment:
Should this / can this be checked here and program aborted if conditions aren't met?

Reply:
I don't prefer aborting the program when we encounter a setting we don't support yet, because I routinely check the performance with under such settings. For example, there are bits of the code where multi-rank is not supported (or bits that are not implemented), so running the solver would definitely give a wrong result, but it is sometimes perfectly representative in terms of performance. This is helping me catching performance related stuff early on and avoiding potential suprises when we run a feature first time after supporting it.

I avoid TODOs as much as I can but here the point is very subtle so I wanted to highlight it at least.

Reply:
In this case, it seems fairly limiting, and I'd argue it would require more than a "TODO" comment. If we aren't aborting, having it at least print to the output would be good. 
Once we reach 'production', I think we should reassess this, as I agree that aborting would be better.

Reply:
In the production code yes of course, aborting in case of incompatible settings is definitely the solution. We can't expect users to know what we support and set things accordingly.

We're now printing that waves array is correct only for a single rank run.